Provide a peek goal to easily view BUILD metadata from command line

jriddy opened the pull request:
### Problem

It would be convenient to have pants able to export build metadata, either for human consumption or programmatic consumption. This seems to be the essence of issue #4861, which the PR addresses.

### Solution

This PR adds a new `peek` goal, which allows us to look at the BUILD files for given targets.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,146 @@
+        """Return a serializable object for o."""
+        if is_dataclass(o):
+            return asdict(o)
+        elif isinstance(o, Mapping):
```
Another common one is `Sequence`.

Also, I think this `isinstance` check is supposed to use `collections.abc.Mapping`, rather than `typing`. To avoid ambiguity, I recommend using the fully specified `collections.abc.Mapping` instead of `from collections.abc import Mapping`.

jriddy replied:
- [ ] add `Sequence` and switch to `collections.abc`

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,146 @@
+    # TODO: better options for target selection (globs, transitive, etc)?
+    targets: Iterable[Target]
+    # TODO: handle target not found exceptions better here
+    targets = await Get(Targets, Addresses, addresses)
```
cc @stuhood on this to confirm.

`await Get(Targets, Addresses)` expands out all "file targets" every time. Meaning, it generates a more granular target which only has one source file, rather than the original "BUILD target" you see in the BUILD file. Run `./pants peek --output=json src/python/pants/util:util` - surprisingly, you'll get >1 target.

Almost certainly, we instead want to use `await Get(UnexpandedTargets, Addresses, addresses)`. This will only give back the precise target specified by the user, i.e. `src/python/pants/util:util` gives you that BUILD target, and then `src/python/pants/util/strutil.py` gives you the more granular file target `strutil.py`.

@stuhood can you think of any scenario here where we wouldn't want to use `UnexapandedTargets`?

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,146 @@
+        build_file_paths = {a.rel_path for a in build_file_addresses}
+        digest_contents = await Get(DigestContents, PathGlobs(build_file_paths))
+        # TODO: programmatically determine correct encoding
+        rendereds = map(_render_raw, digest_contents)
+        for output in _intersperse(os.linesep, rendereds):
```
I'm not following what's happening with the intersperse method? Would it work to instead do this?

```python
for file_content in digest_contents:
  console.print_stdout(_render_raw(file_content))
```

Eric-Arellano replied:
Regarding ordering, I think we likely would want to use alphabetical ordering based on the `file_content.path`.

jriddy replied:
I was trying to get too cute with `intersperse`, I'll simplify.

- [ ] remove `_intersperse()`
- [ ] alphabetically order along paths

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,146 @@
+    JSON = "json"
+
+
+class PeekSubsystem(GoalSubsystem):
```
We'll want to use the `Outputting` mixin so that the `--output-file` option is added: https://www.pantsbuild.org/v2.2/docs/rules-api-goal-rules#outputting-mixin-optional.

jriddy replied:
Okay, seems simple enough

- [ ] add `Outputting` mixin

Eric-Arellano reviewed (commented):
This is awesome, really cool work!

--

For raw output, I agree with the decision to simply print the whole BUILD file. Way too tricky to otherwise try to identify which part is applicable. And we hope/expect BUILD files to be fairly small thanks to dep inference. 

However, I could see it getting confusing when you specify multiple targets/files, and mapping what belongs to what. Thoughts on detecting the case of >1 "spec", and logging a warning to consider using the JSON format? We do that with `./pants py-constraints`:

https://github.com/pantsbuild/pants/blob/50b3e525f4a196773d6f3e3e598b5a08bdbb7224/src/python/pants/backend/python/mixed_interpreter_constraints/py_constraints.py#L173-L180

--

The JSON output is excellent: it shows the user what Pants's view of a particular file/target actually is. That transparency is great for debugging. We might want to make JSON the default?

I'd recommend two tweaks:
1) Rather than `alias`, call it `target_type`, or something similar. "Alias" is an implementation detail.
2) I think we want to render default values? That way, we show a complete view of what Pants is doing. It also allows for safer JSON queries, as you're guaranteed the key will exist. If it gets too chatty, we could consider an option like `--ignore-defaults`.

--

In https://github.com/pantsbuild/pants/issues/4861, we mentioned adding a table format in a followup. I'm not sure that would really be necessary - the mix of Raw and JSON is very helpful already.

--

In https://github.com/pantsbuild/pants/issues/4861, I pondered

> Should [we] evaluate the sources and dependencies fields, or show the raw values?

After seeing this PR, I think no, we should not. `./pants dependencies` already exists for deps, and `./pants filedeps` for sources. It's redundant imo, and also non-trivial, to include those here, including a performance hit.

Eric-Arellano commented:
If you have a chance, it'd also be great to add some tests. You'd want to use Approach #3: Rule Runner, maybe also some normal unit tests (Approach #1).

Here's something to get you started:

```python
# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
# Licensed under the Apache License, Version 2.0 (see LICENSE).

import pytest
from textwrap import dedent

from pants.core.target_types import Files
from pants.backend.project_info import peek
from pants.backend.project_info.peek import Peek
from pants.testutil.rule_runner import RuleRunner


@pytest.fixture
def rule_runner() -> RuleRunner:
    return RuleRunner(rules=peek.rules(), target_types=[Files])


def test_example(rule_runner: RuleRunner) -> None:
    rule_runner.add_to_build_file("project", "# A comment\nfiles(sources=[])")
    result = rule_runner.run_goal_rule(Peek, args=["project"])
    assert result.stdout == dedent(
        """\
        -------------
        project/BUILD
        -------------
        # A comment
        files(sources=[])

        """
    )
```

Note that we likely do want to do exact string matches, as formatting matters for Pants goals and we don't want to regress.

It would be good to test running on both a BUILD target address (like `project:app`) and a file address, like `project/app.py` or `project/app.py:lib`.

jriddy reviewed (commented):
> This is awesome, really cool work!

Thanks! Glad to finally be able to contribute!

> For raw output, I agree with the decision to simply print the whole BUILD file. Way too tricky to otherwise try to identify which part is applicable. And we hope/expect BUILD files to be fairly small thanks to dep inference.

I played (a very small bit) with some inspection tools to find the applicable part, but it looked complicated. It could be an issue in its own right if we wanted to pursue it more.
 
> However, I could see it getting confusing when you specify multiple targets/files, and mapping what belongs to what. Thoughts on detecting the case of >1 "spec", and logging a warning to consider using the JSON format? We do that with `./pants py-constraints`:

I don't think a warning is necessary if we make JSON the default output, but I'll defer to your judgement on this.

> The JSON output is excellent: it shows the user what Pants's view of a particular file/target actually is. That transparency is great for debugging. We might want to make JSON the default?
> 
> I'd recommend two tweaks:
> 
>     1. Rather than `alias`, call it `target_type`, or something similar. "Alias" is an implementation detail.
> 
>     2. I think we want to render default values? That way, we show a complete view of what Pants is doing. It also allows for safer JSON queries, as you're guaranteed the key will exist. If it gets too chatty, we could consider an option like `--ignore-defaults`.
> 

I agree with these suggestions.

- [ ] Make JSON the default output style
- [x] Rename `alias` in output to `target_type`
- [ ] Show default values by default
- [ ] Add `--exclude-defaults` flag (optional)

> In #4861, we mentioned adding a table format in a followup. I'm not sure that would really be necessary - the mix of Raw and JSON is very helpful already.

Agreed. It's human-readable enough, and machine readable with things like `jq`

> In #4861, I pondered
> 
> > Should [we] evaluate the sources and dependencies fields, or show the raw values?
> 
> After seeing this PR, I think no, we should not. `./pants dependencies` already exists for deps, and `./pants filedeps` for sources. It's redundant imo, and also non-trivial, to include those here, including a performance hit.

Agreed, and if you really want that, `jq` + `xargs` are your friends:

```
./pants peek ... | jq -r ... | xargs ./pants filedeps
```

Although I admit that trying to come up with a good example `jq` query made me realize that it doesn't seem to play well with arbitrary keys like we're using, so perhaps making a list of objects is a better way forward. Any opinion on this, @Eric-Arellano ?

jriddy commented:
> If you have a chance, it'd also be great to add some tests. You'd want to use Approach #3: Rule Runner, maybe also some normal unit tests (Approach #1).

I always intended to add tests, just didn't get around to it on the first pass, as I wanted to make sure the behavior was right before I started baking it in. Thanks for the template.

- [ ] unit tests
- [ ] rule runner tests

jsirois commented:
> Although I admit that trying to come up with a good example jq query made me realize that it doesn't seem to play well with arbitrary keys like we're using, ...

FWIW jq supports arbitrary keys by quoting them (you can also index an object with a quoted key - `.["..."]`) so I don't think jq needs to be a strong constraint on the output format you choose:
```
$ echo '
{"foo": "bar"}
{"//super/strange/pants:key": {"baz": 42}}
{"eggs": "spam"}
' | jq '."//super/strange/pants:key".baz'
null
42
null
```

jriddy commented:
I've addressed most of these issues, except for the tests, which I'll start adding.

I do have one question left, which is the encoding of build files. I get them as bytes in the digest, but how do I know what encoding to use to decode those bytes back to a string. The encoding could theoretically vary depending on how the file was saved right? Or do you normalize to a particular encoding like utf-8 when converting to a digest?

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+def _render_raw(fcs: Iterable[FileContent]) -> str:
+    sorted_fcs = sorted(fcs, key=lambda fc: fc.path)
+    rendereds = map(_render_raw_build_file, sorted_fcs)
+    return os.linesep.join(rendereds)
```
Normally we assume `"\n"` given that we don't support Windows (yet?), but this seems fine to use the more general `os.linesep`. That's good to know it exists.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+    return os.linesep.join(rendereds)
+
+
+def _render_raw_build_file(fc: FileContent, encoding: str = "utf-8") -> str:
```
Regarding encoding, we've been assuming UTF-8. We validate the locale in https://github.com/pantsbuild/pants/blob/1b47b30d7d679505f13a9ec3d38d6381284733f3/src/python/pants/bin/pants_loader.py#L45-L54

We assume UTF-8 for BUILD files by leaving off the `encoding` arg:

https://github.com/pantsbuild/pants/blob/1b47b30d7d679505f13a9ec3d38d6381284733f3/src/python/pants/engine/internals/build_files.py#L108-L111

While this certainly is not 100% safe, it's worked for the past few years as far as we can tell. I imagine we will need to revisit this if/when adding Windows support, as I think it uses UTF-16. But that's out of scope for your PR - we'll need to revisit a lot of things for that to work.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+        **{
+            k.alias: v.value
+            for k, v in t.field_values.items()
+            # don't report default values in normalized output
```
Nit: comments should be sentences: https://www.pantsbuild.org/docs/style-guide#style.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+_nothing = object()
+
+
+def _target_to_dict(t: Target, exclude_defaults: bool = False, _nothing: object = _nothing) -> dict:
```
Consider inlining this as a helper function within `_render_json`. You'd be able to reuse the `exclude_defaults` without needing to pass it again, and you'll get more proximity. This is fine as is, though, if you prefer it.

Either way, no need for the `_nothing` arg.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+
+    def default(self, o):
+        """Return a serializable object for o."""
+        if is_dataclass(o):
+            return asdict(o)
+        elif isinstance(o, collections.abc.Mapping):
+            return dict(o)
+        elif isinstance(o, collections.abc.Sequence):
+            return list(o)
+        elif isinstance(o, self.safe_to_str_types):
+            return str(o)
+        else:
+            try:
+                return super().default(o)
+            except TypeError:
+                # punt and just try str()
+                # TODO: can we find a better strategy here?
+                return str(o)
```
Nit: you can leverage early returns to reduce nesting:

```python
        if is_dataclass(o):
            return asdict(o)
        if isinstance(o, collections.abc.Mapping):
            return dict(o)
        if isinstance(o, collections.abc.Sequence):
            return list(o)
        if isinstance(o, self.safe_to_str_types):
            return str(o)
        try:
            return super().default(o)
        except TypeError:
            # punt and just try str()
            # TODO: can we find a better strategy here?
            return str(o)
```

While not required, that style is encouraged: https://www.pantsbuild.org/docs/style-guide#prefer-early-returns-in-functions.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+            return dict(o)
+        elif isinstance(o, collections.abc.Sequence):
+            return list(o)
+        elif isinstance(o, self.safe_to_str_types):
+            return str(o)
```
If you were to remove this, I think it would still end up using `str(o)` in the try block, right? I'm wondering if it's necessary to maintain. It's a bit of a leaky abstraction that `backend.backend.project_info` has any notion of `pants.backend.python` / Python requirements.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+                return super().default(o)
+            except TypeError:
+                # punt and just try str()
+                # TODO: can we find a better strategy here?
```
This is the best I could think of. I think it's reasonable.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+async def peek(
+    console: Console,
+    subsys: PeekSubsystem,
+    addresses: Addresses,
```
Nit: you can move remove 139 by putting `targets: UnexpandedTargets` here. That will cause the engine to go from `Specs -> Addresses -> UnexapandedTargets` for you. 

It's slightly faster as you get more parallelism. As is, things are blocked until the `await Get()` is evaluated, whereas all 3 of those rule inputs can be evaluated in parallel.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+    addresses: Addresses,
+) -> Peek:
+    # TODO: better options for target selection (globs, transitive, etc)?
+    targets: Iterable[Target]
```
No need for this - I suspect bad copy pasta from other goals where `targets` is sometimes `TransitiveTargets.clsoure` and sometimes `Targets`, which are both `Iterable[Target]`. Here, it's only ever `UnexpandedTargets`, which is really a `Collection[Target]`, which is an `Iterable[Target]`.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+    subsys: PeekSubsystem,
+    addresses: Addresses,
+) -> Peek:
+    # TODO: better options for target selection (globs, transitive, etc)?
```
What do you mean? Likely related to https://github.com/pantsbuild/pants/issues/10087 and https://github.com/pantsbuild/pants/issues/7346.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+) -> Peek:
+    # TODO: better options for target selection (globs, transitive, etc)?
+    targets: Iterable[Target]
+    # TODO: handle target not found exceptions better here
```
What do you mean? This should be handled already. I'd be very interested in suggestions for how to improve that error message though - input is always appreciated on error messages.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+        )
+        build_file_paths = {a.rel_path for a in build_file_addresses}
+        digest_contents = await Get(DigestContents, PathGlobs(build_file_paths))
+        # TODO: programmatically determine correct encoding
```
See above thoughts on encoding.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+        output = _render_raw(digest_contents)
+    elif subsys.output_type == OutputOptions.JSON:
+        output = _render_json(targets, subsys.exclude_defaults)
+
```
Might want to add `else: raise AssertionError("...")`.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+            "--output",
+            type=OutputOptions,
+            default=OutputOptions.JSON,
+            help="which output style peek should use",
```
Should be capitalized. Also would be helpful to mention a bit more, maybe something like this:

> `json` will show each target as a separate entry, whereas `raw` will simply show the original BUILD file(s) the input files/targets come from.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+            "--exclude-defaults",
+            type=bool,
+            default=False,
+            help="whether to exclude values matching default field values from normalized output",
```
Ditto on capitalizing. 

With docs, imo, it can sometimes be helpful to say the same thing in a different way. My thinking is if the person is not familiar with that word, such as English language learners. Rather than repeating "exclude", you can use something like "Whether to leave off values that match Pants's default values when using `json` output."

Eric-Arellano reviewed (commented):
Awesome, looking great!

jriddy commented:
Thanks for the feedback. I'll update according to the comments and add some real test cases, but I won't have time to do this until about the 24th or so. Since it's the holidays, I won't expect responses. Have a good one!

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,159 @@
+
+
+class PeekSubsystem(Outputting, GoalSubsystem):
+    """Display BUILD file info to the console.
+
+    In its most basic form, `peek` just prints the contents of a BUILD file. It can also display
+    multiple BUILD files, or render normalized target metadata as JSON for consumption by other
+    programs.
+    """
```
As of yesterday, this should now be set through a class property `help`, e.g.

https://github.com/pantsbuild/pants/blob/7d8228330c77b8bd35c3411ae8375da838bdbf33/src/python/pants/vcs/changed.py#L89-L94

Note the newline characters - we want the first paragraph to still be only one sentence like this.

You can test that both `./pants help goals` and `./pants help peek` render correctly.

jriddy commented:
Sorry I left this dormant so long. I mostly just forgot about it (or had it at the very back of my mind) until @cczona pinged me to remind me about this, so thanks for that!

I've addressed the previous comments and added some tests. I couldn't find a much better way to test it than to just test raw output generation. although I could come up with some more scenarios with a generator maybe, but y'all don't seem to be using `hypothesis` or another generative testing tool, so I guess I'll leave it at some hand-written examples.

Let me know if there's anything left I need to do to make this cut the mustard.

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,157 @@
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
```
```suggestion
# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
```

Eric-Arellano commented on src/python/pants/backend/project_info/peek.py:
```diff
@@ -0,0 +1,157 @@
+async def peek(
+    console: Console,
+    subsys: PeekSubsystem,
+    addresses: Addresses,
+) -> Peek:
+    targets: Iterable[Target]
+    targets = await Get(UnexpandedTargets, Addresses, addresses)
```
Bump that you can do this:

```diff
diff --git a/src/python/pants/backend/project_info/peek.py b/src/python/pants/backend/project_info/peek.py
index e8f9dbf92..09b324fb6 100644
--- a/src/python/pants/backend/project_info/peek.py
+++ b/src/python/pants/backend/project_info/peek.py
@@ -10,7 +10,7 @@ from typing import Any, Iterable, Mapping, cast
 
 from pkg_resources import Requirement
 
-from pants.engine.addresses import Address, Addresses, BuildFileAddress
+from pants.engine.addresses import Address, BuildFileAddress
 from pants.engine.console import Console
 from pants.engine.fs import DigestContents, FileContent, PathGlobs
 from pants.engine.goal import Goal, GoalSubsystem, Outputting
@@ -130,11 +130,8 @@ class _PeekJsonEncoder(json.JSONEncoder):
 async def peek(
     console: Console,
     subsys: PeekSubsystem,
-    addresses: Addresses,
+    targets: UnexpandedTargets,
 ) -> Peek:
-    targets: Iterable[Target]
-    targets = await Get(UnexpandedTargets, Addresses, addresses)
-
     if subsys.output_type == OutputOptions.RAW:
         build_file_addresses = await MultiGet(
             Get(BuildFileAddress, Address, t.address) for t in targets
```

Eric-Arellano reviewed (approved):
Great stuff! One small change

Eric-Arellano commented on src/python/pants/backend/project_info/peek_test.py:
```diff
@@ -0,0 +1,160 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
```
```suggestion
# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
```

stuhood reviewed (approved):
Looks great: thanks a lot for iterating on this!
